### PR TITLE
Draft of Signup and Login Usecase

### PR DESCRIPTION
**Pull request by modaserMoj**
This PR contains all the code for the usecase layer of the login and signup feature.

### CODE REVIEW

File: src/main/java/ca/utoronto/fitbook/adapter/web/UserLoginController.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+
+@Controller
+@RequiredArgsConstructor
+public class UserLoginController {
+    private final UserLoginUseCase userLoginUseCase;
+
+    @PostMapping(path = "/post")
+    String createPost(Model model, @RequestBody UserLoginCommand command) {
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/web/UserLoginController.java:21]: This should map to /login and the method should be called userLogin() or something.

File: src/main/java/ca/utoronto/fitbook/adapter/web/UserRegisterController.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+
+@Controller
+@RequiredArgsConstructor
+public class UserRegisterController {
+    private final UserRegisterUseCase userRegisterUseCase;
+
+    @PostMapping(path = "/post")
+    String createPost(Model model, @RequestBody UserRegisterCommand command) {
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/adapter/web/UserRegisterController.java:18]: ditto

File: src/main/java/ca/utoronto/fitbook/application/port/out/response/UserLoginResponse.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package ca.utoronto.fitbook.application.port.out.response;
+import lombok.Value;
+
+@Value
+public class UserLoginResponse {
+
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/port/out/response/UserLoginResponse.java:6]: Extra space

File: src/main/java/ca/utoronto/fitbook/application/service/UserLoginService.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    private final LoadUserByNamePort loadUserByNamePort;
+    private final FindUserByNamePort findUserByNamePort;
+
+    @Override
+    public UserLoginResponse loginUser(UserLoginCommand command){
+        if(!findUserByNamePort.findByName(command.getName())){
+
+            throw new UserNotFound();
+
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/UserLoginService.java:27]: Extra spaces

File: src/main/java/ca/utoronto/fitbook/application/service/UserLoginService.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        } else if (loadUserByNamePort.loadUser(command.getName()).getPassword() != command.getPassword()){
+
+            throw new IncorrectPassword();
+        }
+        User user = loadUserByNamePort.loadUser(command.getName());
+        return new UserLoginResponse(user.getId());
+    }
+    @ResponseStatus(value= HttpStatus.INTERNAL_SERVER_ERROR, reason="Incorrect Password")
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/UserLoginService.java:35]: Find a better status code maybe, a 401 Unauthorized for wrong password

File: src/main/java/ca/utoronto/fitbook/application/service/UserLoginService.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    public class IncorrectPassword extends RuntimeException {
+        public IncorrectPassword() {
+            super("Incorrect Password.");
+        }
+    }
+    @ResponseStatus(value= HttpStatus.INTERNAL_SERVER_ERROR, reason="User Not Found")
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/UserLoginService.java:41]: User not found should probably be a 422 Unprocessable Entity

**Review by youssefsoli — CHANGES_REQUESTED** [review]
The structure and logic look good, there are just a few small style issues and nit picks.

**Comment by modaserMoj** [comment]
The reviewer requested changes were made

**Review by jpablo2002 — APPROVED** [review]
Code is written cleanly and the creation of custom exceptions is great to better represent the cause of the errors! Looks good to me!

File: src/main/java/ca/utoronto/fitbook/application/port/out/response/UserRegisterResponse.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package ca.utoronto.fitbook.application.port.out.response;
+
+import lombok.Value;
+
+@Value
+public class UserRegisterResponse
+{
+    String id;
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/port/out/response/UserRegisterResponse.java:8]: ditto

File: src/main/java/ca/utoronto/fitbook/application/service/UserRegisterService.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    @ResponseStatus(value= HttpStatus.NOT_ACCEPTABLE, reason="Password don't match")
+    public class PasswordNotMatch extends RuntimeException {
+        public PasswordNotMatch() {
+            super("Password don't match.");
+        }
+    }
+    @ResponseStatus(value= HttpStatus.NOT_ACCEPTABLE, reason="Name is too long")
+    public class NameTooLong extends RuntimeException {
+        public NameTooLong(UserRegisterCommand command) {
+            super("Name is too long by "+ ((command.getName().length())-40)+" characters");
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/UserRegisterService.java:64]: Add space between operators

File: src/main/java/ca/utoronto/fitbook/application/service/UserRegisterService.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+    @ResponseStatus(value= HttpStatus.NOT_ACCEPTABLE, reason="Password is too long")
+    public class PasswordTooLong extends RuntimeException {
+        public PasswordTooLong(UserRegisterCommand command) {
+            super("Password is too long by "+ ((command.getName().length())-40)+" characters");
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/UserRegisterService.java:70]: ditto

**Review by youssefsoli — CHANGES_REQUESTED** [review]

**Comment by jannater** [comment]
Good job with clean architecture and  hexagonal architecture while implementing the use cases.

File: src/main/java/ca/utoronto/fitbook/application/service/UserRegisterService.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            throw  new UsernameAlreadyExists();
+        } else if (!command.getPassword().equals(command.getRepeatedPassword())) {
+            throw new PasswordNotMatch();
+        } else if (command.getName().length() > 40){
+            throw new NameTooLong(command);
+        } else if (command.getPassword().length() < 8){
+            throw new PasswordTooShort();
+        } else if (command.getPassword().length() > 40){
+            throw new PasswordTooLong(command);
+        }
```
Comment (youssefsoli) [inline src/main/java/ca/utoronto/fitbook/application/service/UserRegisterService.java:36]:
```suggestion
        if (!command.getPassword().equals(command.getRepeatedPassword()))
            throw new PasswordNotMatch();
         if (command.getName().length() > 40)
            throw new NameTooLong(command);
        if (command.getPassword().length() < 8)
            throw new PasswordTooShort();
        if (command.getPassword().length() > 40)
            throw new PasswordTooLong(command);
```

**Review by Ludosss — APPROVED** [review]
looks good!

**Review by jpablo2002 — APPROVED** [review]
Looks good! Great fixes!

**Pull request merged**